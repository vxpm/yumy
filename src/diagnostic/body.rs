mod builder;
mod writer;

use crate::{source::SourceLine, Config, Label, Source};

/// A chunk of a diagnostic's body.
#[derive(Debug)]
pub(super) struct BodyChunk<'src> {
    pub line: SourceLine<'src>,
    pub singleline_labels: Vec<Label>,
    pub starting_multiline_labels: Vec<Label>,
    pub finishing_multiline_labels: Vec<usize>,
}

/// Describes a body and contains some cached useful information about it.
#[derive(Debug)]
pub(super) struct BodyDescriptor<'src> {
    /// The chunks of this body.
    pub chunks: Vec<BodyChunk<'src>>,
    /// How much indentation can be trimmed off in every line.
    pub indent_trim: usize,
    /// The width needed to display all line numbers in the body.
    pub line_number_width: usize,
    /// The maximum number of parallel labels that happen in the body.
    pub maximum_parallel_labels: usize,
}

impl<'src> BodyDescriptor<'src> {
    /// Builds a new [`BodyDescriptor`].
    pub(super) fn new(source: Source<'src>, labels: Vec<Label>) -> Self {
        builder::DescriptorBuilder::new(source, labels).build()
    }

    /// Writes the body described by this descriptor to a given writer.
    pub(super) fn write_to<W>(self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: std::io::Write,
    {
        writer::BodyWriter::new(writer, config.clone(), self).write()
    }
}
