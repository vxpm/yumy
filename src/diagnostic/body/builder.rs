use super::{BodyChunk, BodyDescriptor};
use crate::{source::Source, Label, SourceSpan};

/// Struct that takes care of building a descriptor.
/// Keeping the state for this in it's own struct is easier.
pub(super) struct DescriptorBuilder<'src> {
    source: Source<'src>,
    labels: Vec<Label>,
    active_labels: Vec<(usize, SourceSpan)>,
    multiline_id: usize,
    current_line: usize,
    indent_trim: usize,
    result: Vec<BodyChunk<'src>>,
}

impl<'src> DescriptorBuilder<'src> {
    /// Creates a new [`DescriptorBuilder`].
    pub(super) fn new(source: Source<'src>, mut labels: Vec<Label>) -> Self {
        // sort labels by their start, in reverse order (make it a stack)
        labels.sort_by_key(|label| std::cmp::Reverse(label.span.start()));
        Self {
            source,
            labels,
            active_labels: Vec::new(),
            multiline_id: 0,
            current_line: 0,
            indent_trim: usize::MAX,
            result: Vec::new(),
        }
    }

    /// Returns the singleline labels of the current line and the multiline labels starting in the
    /// current line, respectively.
    fn emit_labels_in_current(&mut self) -> (Vec<Label>, Vec<Label>) {
        let mut singleline_labels = Vec::new();
        let mut multiline_labels = Vec::new();
        loop {
            let Some(label) = self.labels.pop() else {
                break;
            };

            let label_start_line = self
                .source
                .line_index_at(label.span.start() as usize)
                .expect("valid label");

            if label_start_line == self.current_line {
                if label.is_singleline(&self.source) {
                    singleline_labels.push(label);
                } else {
                    self.active_labels.push((self.multiline_id, label.span));
                    multiline_labels.push(label);
                    self.multiline_id += 1;
                }
            } else {
                // put the label back
                self.labels.push(label);
                break;
            }
        }

        (singleline_labels, multiline_labels)
    }

    /// Returns the ID of multiline labels finishing in the current line.
    fn finish_labels_in_current(&mut self) -> Vec<usize> {
        let mut finished_multiline_labels = Vec::new();
        self.active_labels.retain(|(label_id, span)| {
            let finished = self
                .source
                .line_index_at(span.end() as usize)
                .expect("valid label")
                == self.current_line;

            if finished {
                finished_multiline_labels.push(*label_id);
            }

            !finished
        });

        finished_multiline_labels
    }

    fn emit_events(&mut self) {
        // here's how it should go:
        // - if no active labels:
        // -- find next label and jump to its start
        // -- if singleline, emit the label
        // -- if multiline, start it
        // - if active labels:
        // -- go line by line
        // -- if a singleline label is in its start, emit it
        // -- if a multiline ends, remove it

        while !self.labels.is_empty() || !self.active_labels.is_empty() {
            if self.active_labels.is_empty() {
                let label = self.labels.last().expect("has remaining labels");
                self.current_line = self
                    .source
                    .line_index_at(label.span.start() as usize)
                    .expect("label span is valid");
            } else {
                self.current_line += 1;
            }

            let line = self
                .source
                .lines()
                .get(self.current_line)
                .copied()
                .expect("valid line");

            // special case: if the line is empty, don't consider it for indent trimming
            if !line.text().is_empty() {
                self.indent_trim = self.indent_trim.min(line.indent_size());
            }

            let (singleline_labels, starting_multiline_labels) = self.emit_labels_in_current();
            let finishing_multiline_labels = self.finish_labels_in_current();
            let chunk = BodyChunk {
                line,
                singleline_labels,
                starting_multiline_labels,
                finishing_multiline_labels,
            };
            self.result.push(chunk);
        }
    }

    /// Calculates the width of the line number section in the body.
    fn calculate_line_number_width(&self) -> usize {
        self.result
            .last()
            .map(|chunk| (chunk.line.index() + 1).ilog10() as usize + 1)
            .unwrap_or(0)
    }

    /// Calculates the maximum number of parallel multiline labels that happens in this descriptor.
    fn calculate_maximum_parallel_labels(&self) -> usize {
        let mut count = 0;
        let mut max = 0;
        for chunk in self.result.iter() {
            count += chunk.starting_multiline_labels.len();
            max = max.max(count);

            // NOTE: this is >after< we recalculate the maximum because labels that finish on a
            // line are still shown on it!
            count -= chunk.finishing_multiline_labels.len();
        }

        max
    }

    /// Builds the [`BodyDescriptor`].
    pub(crate) fn build(mut self) -> BodyDescriptor<'src> {
        self.emit_events();

        let line_number_width = self.calculate_line_number_width();
        let maximum_parallel_labels = self.calculate_maximum_parallel_labels();

        BodyDescriptor {
            chunks: self.result,
            indent_trim: self.indent_trim,
            line_number_width,
            maximum_parallel_labels,
        }
    }
}

#[cfg(test)]
mod test {
    use owo_colors::Style;

    use super::*;
    use crate::{Label, Source};

    #[test]
    fn test_build_singleline() {
        let src = Source::new(crate::test::RUST_SAMPLE_1, Some("src/lib.rs"));
        let labels = vec![
            Label::new(53..66u32, ""),
            Label::new(83..87u32, "recursive without indirection"),
        ];

        crate::test::setup_insta!();
        insta::assert_debug_snapshot!(BodyDescriptor::new(src, labels));
    }

    #[test]
    fn test_build_multiline_1() {
        let src = Source::new(crate::test::RUST_SAMPLE_2, Some("src/main.rs"));
        let labels = vec![
            Label::styled(
                247..260u32,
                "required by a bound introduced by this call",
                Style::new().yellow(),
            ),
            Label::styled(
                261..357u32,
                "`Rc<Mutex<i32>>` cannot be sent between threads safely",
                Style::new().red(),
            ),
        ];

        crate::test::setup_insta!();
        insta::assert_debug_snapshot!(BodyDescriptor::new(src, labels));
    }

    #[test]
    fn test_build_multiline_2() {
        let src = Source::new(crate::test::TEXT_SAMPLE_2, Some("just testing"));
        let labels = vec![
            Label::new(0..36u32, "just testing two multilines"),
            Label::new(10..24u32, "hi"),
            Label::new(28u32..35u32, "hello"),
        ];

        crate::test::setup_insta!();
        insta::assert_debug_snapshot!(BodyDescriptor::new(src, labels));
    }
}
