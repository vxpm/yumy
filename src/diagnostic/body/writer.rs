use super::{BodyChunk, BodyDescriptor, Label};
use crate::Config;
use owo_colors::OwoColorize;
use std::io::Write;

#[derive(Debug, Clone)]
struct ActiveMultiline {
    label_id: usize,
    label: Label,
    recently_added: bool,
}

/// Struct responsible for writing a body described by a [`BodyDescriptor`].
pub(super) struct BodyWriter<'src, W> {
    writer: W,
    config: Config,
    descriptor: BodyDescriptor<'src>,
    slots: Vec<Option<ActiveMultiline>>,
    multiline_id: usize,
    line_number_width: usize,
    current_indent_level: usize,
}

impl<'src, W> BodyWriter<'src, W>
where
    W: Write,
{
    pub(super) fn new(writer: W, config: Config, descriptor: BodyDescriptor<'src>) -> Self {
        let slots_needed = descriptor.maximum_parallel_labels;
        let line_number_width = descriptor.line_number_width;

        Self {
            writer,
            config,
            descriptor,
            slots: vec![None; slots_needed],
            multiline_id: 0,
            line_number_width,
            current_indent_level: 0,
        }
    }

    fn emit_left_column(&mut self, line_index: Option<usize>) -> std::io::Result<()> {
        if let Some(index) = line_index {
            write!(
                self.writer,
                "{:padding$} {} ",
                (index + 1).style(self.config.styles.left_column),
                self.config
                    .charset
                    .vertical_bar
                    .style(self.config.styles.left_column),
                padding = self.line_number_width
            )?;
        } else {
            write!(
                self.writer,
                "{:padding$} {} ",
                "",
                self.config
                    .charset
                    .separator
                    .style(self.config.styles.left_column),
                padding = self.line_number_width
            )?;
        }

        Ok(())
    }

    fn emit_multiline_indicators(
        &mut self,
        finishing_multiline_labels: &[usize],
    ) -> std::io::Result<()> {
        for slot in self.slots.iter_mut() {
            let Some(slot) = slot else {
                write!(self.writer, "  ")?;
                continue;
            };

            let style = slot
                .label
                .indicator_style
                .unwrap_or(self.config.styles.multiline_indicator);

            let indicator_char = if slot.recently_added {
                self.config.charset.multiline_start
            } else if finishing_multiline_labels.contains(&slot.label_id) {
                self.config.charset.multiline_end
            } else {
                self.config.charset.vertical_bar
            };

            slot.recently_added = false;
            write!(self.writer, "{} ", indicator_char.style(style))?;
        }

        Ok(())
    }

    fn emit_source_line(&mut self, chunk: &BodyChunk) -> std::io::Result<()> {
        self.emit_left_column(Some(chunk.line.index()))?;
        self.emit_multiline_indicators(&chunk.finishing_multiline_labels)?;

        // remember the special case: if the line is empty, don't
        // attempt to trim it
        self.current_indent_level = if chunk.line.text().is_empty() {
            0
        } else {
            chunk.line.indent_size() - self.descriptor.indent_trim
        };

        // finally, write the line
        writeln!(
            self.writer,
            "{:l$}{}",
            "",
            chunk.line.text().style(self.config.styles.source),
            l = self.current_indent_level,
        )?;
        Ok(())
    }

    fn emit_singleline_labels(&mut self, chunk: &mut BodyChunk) -> std::io::Result<()> {
        let line = chunk.line;
        let labels = std::mem::take(&mut chunk.singleline_labels);
        for label in labels {
            self.emit_left_column(None)?;
            self.emit_multiline_indicators(&chunk.finishing_multiline_labels)?;

            // calculate ranges into the line text
            let local_base = line.dedented_span().start();
            let before_underline_range = 0usize..(label.span.start() - local_base) as usize;
            let underline_range = before_underline_range.end
                ..(label.span.end().min(line.text().len() as u32) - local_base) as usize;

            // compute widths
            let before_underline_width =
                crate::text::dislay_width(&line.text()[before_underline_range]);
            let underline_width = crate::text::dislay_width(&line.text()[underline_range]);

            // write label
            let before_underline = std::iter::repeat(' ').take(before_underline_width);
            let underline = std::iter::repeat(self.config.charset.underliner).take(underline_width);
            let before_label: String = before_underline.chain(underline).collect();
            let label_style = label
                .indicator_style
                .unwrap_or(self.config.styles.singleline_indicator);

            writeln!(
                self.writer,
                "{:l$}{} {}",
                "",
                before_label.style(label_style),
                label.message.style(label_style),
                l = self.current_indent_level,
            )?;
        }

        Ok(())
    }

    fn allocate_multiline(&mut self, label: Label) {
        let slot = self
            .slots
            .iter_mut()
            .find(|slot| slot.is_none())
            .expect("has enough slots");

        *slot = Some(ActiveMultiline {
            label_id: self.multiline_id,
            label,
            recently_added: true,
        });
        self.multiline_id += 1;
    }

    fn start_multiline_labels(&mut self, chunk: &mut BodyChunk) -> std::io::Result<()> {
        let labels = std::mem::take(&mut chunk.starting_multiline_labels);
        for label in labels {
            self.allocate_multiline(label);
        }

        Ok(())
    }

    fn emit_multiline_label(&mut self, label_id: usize) -> std::io::Result<()> {
        self.emit_left_column(None)?;

        let mut finished_multiline = None;
        let mut slots_iter = self.slots.iter_mut();
        while let Some(slot) = slots_iter.next() {
            let Some(active) = slot else {
                write!(self.writer, "  ",)?;
                continue;
            };

            let style = active
                .label
                .indicator_style
                .unwrap_or(self.config.styles.multiline_indicator);

            if active.label_id == label_id {
                finished_multiline = std::mem::take(slot);
                write!(
                    self.writer,
                    "{}{}",
                    self.config.charset.connection_top_to_right.style(style),
                    self.config.charset.horizontal_bar.style(style)
                )?;
                break;
            }

            write!(
                self.writer,
                "{} ",
                self.config.charset.vertical_bar.style(style)
            )?;
        }

        let finished_label = finished_multiline.unwrap().label;
        let finished_style = finished_label
            .indicator_style
            .unwrap_or(self.config.styles.multiline_indicator);

        while let Some(slot) = slots_iter.next() {
            let Some(active) = slot else {
                write!(self.writer, "  ",)?;
                continue;
            };

            let style = active
                .label
                .indicator_style
                .unwrap_or(self.config.styles.multiline_indicator);

            write!(
                self.writer,
                "{}{}",
                self.config.charset.multiline_crossing.style(style),
                self.config.charset.horizontal_bar.style(finished_style)
            )?;
        }

        writeln!(
            self.writer,
            " {}",
            finished_label.message.style(finished_style)
        )?;
        Ok(())
    }

    fn finish_multiline_labels(&mut self, chunk: &mut BodyChunk) -> std::io::Result<()> {
        let labels = std::mem::take(&mut chunk.finishing_multiline_labels);
        for label_id in labels {
            self.emit_multiline_label(label_id)?;
        }

        Ok(())
    }

    pub(super) fn write(mut self) -> std::io::Result<()> {
        let chunks = std::mem::take(&mut self.descriptor.chunks);
        for mut chunk in chunks {
            self.start_multiline_labels(&mut chunk)?;
            self.emit_source_line(&chunk)?;
            self.emit_singleline_labels(&mut chunk)?;
            self.finish_multiline_labels(&mut chunk)?;
        }

        Ok(())
    }
}
