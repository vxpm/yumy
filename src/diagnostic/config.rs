use owo_colors::Style;

/// The charset to use when rendering a diagnostic.
#[derive(Debug, Clone)]
pub struct Charset {
    /// A vertical bar.
    pub vertical_bar: char,
    /// An horizontal bar.
    pub horizontal_bar: char,
    /// The character used to underline the source
    /// in single-line labels.
    pub underliner: char,
    /// The character that's used instead of the vertical
    /// bar when not in a source line.
    pub separator: char,
    /// The character that connects the vertical bar
    /// to the connector in multiline labels.
    pub connection_top_to_right: char,
    /// The character for when a multiline label starts.
    pub multiline_start: char,
    /// The character for when a multiline label ends.
    pub multiline_end: char,
    /// The character for when two multiline labels cross.
    pub multiline_crossing: char,
}

impl Default for Charset {
    fn default() -> Self {
        Self {
            vertical_bar: '│',
            horizontal_bar: '╶',
            underliner: '^',
            separator: ':',
            connection_top_to_right: '╰',
            multiline_start: '┬',
            multiline_end: '┼',
            multiline_crossing: '┼',
        }
    }
}

/// Default styles to use for each part of a diagnostic.
#[derive(Debug, Clone)]
pub struct DefaultStyles {
    pub source_name: Style,
    pub source: Style,
    pub left_column: Style,
    pub multiline_indicator: Style,
    pub singleline_indicator: Style,
    pub footnote_indicator: Style,
}

impl Default for DefaultStyles {
    fn default() -> Self {
        Self {
            source_name: Style::new().white().bold(),
            source: Style::new().white(),
            left_column: Style::new().bright_blue().bold(),
            multiline_indicator: Style::new().yellow(),
            singleline_indicator: Style::new().yellow(),
            footnote_indicator: Style::new().bright_blue().bold(),
        }
    }
}

/// Configuration used to render a diagnostic.
#[derive(Debug, Clone, Default)]
pub struct Config {
    pub charset: Charset,
    pub styles: DefaultStyles,
}
