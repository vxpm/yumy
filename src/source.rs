use nonmax::NonMaxU32;
use owo_colors::Style;

/// Unit struct that represents the absence of
/// a source in a diagnostic.
#[derive(Debug, Clone, Copy)]
pub struct NoSource;

/// A span into the source of a [`crate::Diagnostic`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SourceSpan {
    start: NonMaxU32,
    end: u32,
}

impl SourceSpan {
    /// Creates a new source span. `start` and `end`
    /// are byte indexes into the source.
    pub fn new(start: u32, end: u32) -> Self {
        assert!(end >= start);
        Self {
            start: NonMaxU32::new(start).expect("Start is non-max"),
            end,
        }
    }

    /// The start of this span. Inclusive.
    #[inline]
    pub fn start(&self) -> u32 {
        self.start.get()
    }

    /// The end of this span. Exclusive.
    #[inline]
    pub fn end(&self) -> u32 {
        self.end
    }

    /// The length of this span.
    #[inline]
    pub fn len(&self) -> u32 {
        self.end - self.start()
    }

    /// Does this span contain the given value?
    #[inline]
    pub fn contains(&self, value: u32) -> bool {
        (self.start() <= value) && (value < self.end)
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SourceLine<'src> {
    pub span: SourceSpan,
    pub line: &'src str,
}

impl<'src> SourceLine<'src> {
    pub fn new(line: &'src str, span: SourceSpan) -> Self {
        Self { span, line }
    }
}

/// A source of text to use with a diagnostic.
#[derive(Debug, Clone)]
pub struct Source<'src> {
    src: &'src str,
    name: Option<&'src str>,
    style: Option<Style>,
    lines: Vec<SourceLine<'src>>,
}

impl<'src> Source<'src> {
    fn lines_of(src: &str) -> Vec<SourceLine> {
        let base_addr = src.as_ptr();
        let lines = src.lines().map(|line| {
            let line_addr = line.as_ptr();
            let offset = (line_addr as usize)
                .checked_sub(base_addr as usize)
                .expect("line should always have higher address");
            let end = offset + line.len();

            SourceLine {
                span: SourceSpan::new(offset as u32, end as u32),
                line,
            }
        });

        lines.collect()
    }

    /// Creates a new source.
    pub fn new(src: &'src str, name: Option<&'src str>) -> Self {
        Self {
            src,
            name,
            style: None,
            lines: Self::lines_of(src),
        }
    }

    /// Creates a new source with the given style.
    pub fn styled(src: &'src str, name: Option<&'src str>, style: Style) -> Self {
        Self {
            src,
            name,
            style: Some(style),
            lines: Self::lines_of(src),
        }
    }

    pub fn src(&self) -> &'src str {
        self.src
    }

    pub fn name(&self) -> Option<&'src str> {
        self.name
    }

    pub fn style(&self) -> Option<Style> {
        self.style
    }

    pub fn line(&self, index: u32) -> Option<SourceLine<'src>> {
        self.lines.get(index as usize).copied()
    }

    pub fn lines(&self) -> impl Iterator<Item = &SourceLine> {
        self.lines.iter()
    }

    pub(crate) fn line_index_of_byte(&self, byte_index: u32) -> Option<u32> {
        if byte_index > self.src.len() as u32 {
            return None;
        }

        // TODO: change to a binary search
        for (index, line) in self.lines.iter().enumerate() {
            if line.span.contains(byte_index) {
                return Some(index as u32);
            }

            if line.span.start() > byte_index {
                return Some((index - 1) as u32);
            }
        }

        None
    }
}

#[cfg(test)]
mod test {
    use super::*;

    const SAMPLE: &str = include_str!("../sample.txt");

    #[test]
    fn test_lines() {
        let src = Source::new(SAMPLE, None);
        let mut lines = src.lines();

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(0, 20),
                line: "hello there darling!"
            }),
            lines.next()
        );

        // notice how index 20 is not included in any line
        // since it is the \n character!

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(21, 28),
                line: "this is"
            }),
            lines.next()
        );

        // the same goes for 28

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(29, 29),
                line: ""
            }),
            lines.next()
        );

        // index 29 is the \n! but it is also not included since
        // the line is empty (29..29)

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(30, 46),
                line: "a sample text :)"
            }),
            lines.next()
        );
    }
}
