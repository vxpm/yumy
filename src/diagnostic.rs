mod body;

/// Module for diagnostic configuration related items.
pub mod config;

use self::config::Config;
use super::source::{NoSource, Source, SourceSpan};
use body::BodyWriter;
use owo_colors::{OwoColorize, Style};
use std::{
    io::{BufWriter, Write},
    ops::Range,
};

/// A label is a message that points to a specific
/// part of the source of a [`Diagnostic`].
#[derive(Debug, Clone)]
pub struct Label {
    /// The message of this label.
    pub message: String,
    /// The span this label refers to.
    pub span: SourceSpan,
    /// The indicator style of this label.
    pub indicator_style: Option<Style>,
}

impl Label {
    /// Creates a new label.
    pub fn new<S, M>(span: S, message: M) -> Self
    where
        S: Into<SourceSpan>,
        M: ToString,
    {
        Self {
            message: message.to_string(),
            span: span.into(),
            indicator_style: None,
        }
    }

    /// Creates a new label with the given style for it's indicator.
    pub fn styled<S, M>(span: S, message: M, style: Style) -> Self
    where
        S: Into<SourceSpan>,
        M: ToString,
    {
        Self {
            message: message.to_string(),
            span: span.into(),
            indicator_style: Some(style),
        }
    }

    /// Returns the line range of this label in the given source.
    ///
    /// # Panics
    /// Panics if the span is out of bounds.
    fn line_range(&self, src: &Source) -> Range<u32> {
        src.line_range_of_span(self.span)
            .expect("label should have span in range")
    }

    /// Returns whether this label is singleline or not.
    ///
    /// # Panics
    /// Panics if [`Label::line_range`] would panic.
    fn is_singleline(&self, src: &Source) -> bool {
        let line_range = self.line_range(src);
        line_range.start + 1 == line_range.end
    }
}

/// A diagnostic.
#[derive(Debug, Clone)]
pub struct Diagnostic<Src> {
    message: String,
    labels: Vec<Label>,
    footnotes: Vec<String>,
    source: Src,
}

impl Diagnostic<NoSource> {
    /// Create a new diagnostic without an associated source.
    #[inline]
    pub fn new<M>(message: M) -> Self
    where
        M: ToString,
    {
        Self {
            message: message.to_string(),
            labels: Vec::new(),
            footnotes: Vec::new(),
            source: NoSource,
        }
    }

    /// Attach a source to this diagnostic.
    #[inline(always)]
    pub fn with_source(self, source: Source<'_>) -> Diagnostic<Source<'_>> {
        Diagnostic {
            message: self.message,
            labels: self.labels,
            footnotes: self.footnotes,
            source,
        }
    }
}

impl<Src> Diagnostic<Src> {
    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn with_message<M>(mut self, message: M) -> Self
    where
        M: ToString,
    {
        self.message = message.to_string();
        self
    }

    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn add_label(&mut self, label: Label) {
        self.labels.push(label);
    }

    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn with_label(mut self, label: Label) -> Self {
        self.add_label(label);
        self
    }

    /// Replaces the [`Label`]s of this diagnostic.
    #[inline(always)]
    pub fn with_labels(mut self, labels: Vec<Label>) -> Self {
        self.labels = labels;
        self
    }

    /// Add a footnote to this diagnostic. A footnote is a message
    /// shown after the body of a diagnostic.
    #[inline(always)]
    pub fn add_footnote<F>(&mut self, footnote: F)
    where
        F: ToString,
    {
        self.footnotes.push(footnote.to_string());
    }

    /// Add a footnote to this diagnostic. A footnote is a message
    /// shown after the body of a diagnostic.
    #[inline(always)]
    pub fn with_footnote<F>(mut self, footnote: F) -> Self
    where
        F: ToString,
    {
        self.add_footnote(footnote.to_string());
        self
    }
}

impl<'src> Diagnostic<Source<'src>> {
    /// Calculates the left padding necessary for this diagnostic.
    fn left_padding(&self) -> usize {
        let mut padding = 0;
        for label in &self.labels {
            // find last line of label
            let line_index = self
                .source
                .line_index_at(label.span.end().saturating_sub(1));
            let index_algs = line_index
                .map(|x| f32::log10(x as f32).floor() as usize)
                .unwrap_or(0);
            if index_algs > padding {
                padding = index_algs;
            }
        }

        padding + 1
    }

    fn write_header<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        writeln!(writer, "{}", self.message)?;

        let left_padding = self.left_padding();
        write!(writer, "{:padding$}", "", padding = left_padding)?;
        writeln!(
            writer,
            " {} {}{}{}",
            '@'.style(config.styles.left_column),
            '['.style(config.styles.left_column),
            self.source
                .name()
                .unwrap_or("unknown")
                .style(config.styles.source_name),
            ']'.style(config.styles.left_column)
        )?;
        Ok(())
    }

    fn write_header_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        writeln!(writer, "{}", self.message)?;
        writeln!(
            writer,
            "{} {}{}{}",
            '@'.style(config.styles.left_column),
            '['.style(config.styles.left_column),
            self.source
                .name()
                .unwrap_or("unknown")
                .style(config.styles.source_name),
            "]:".style(config.styles.left_column)
        )?;
        Ok(())
    }

    fn write_body<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        let body_lines = BodyWriter::new(
            writer,
            self.source.clone(),
            config,
            self.left_padding(),
            self.labels.as_slice(),
        );

        body_lines.write()?;

        Ok(())
    }

    fn write_body_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        for label in &self.labels {
            let range = label.line_range(&self.source);
            if range.start + 1 == range.end {
                writeln!(
                    writer,
                    "{} {}{}{}{}{}",
                    config.charset.vertical_bar.style(config.styles.left_column),
                    '['.style(config.styles.left_column),
                    "line ".style(config.styles.source),
                    range.start.style(config.styles.source),
                    "]: ".style(config.styles.left_column),
                    label.message
                )?;
            } else {
                writeln!(
                    writer,
                    "{} {}{}{:?}{}{}",
                    config.charset.vertical_bar.style(config.styles.left_column),
                    '['.style(config.styles.left_column),
                    "lines ".style(config.styles.source),
                    range.style(config.styles.source),
                    "]: ".style(config.styles.left_column),
                    label.message
                )?;
            }
        }

        Ok(())
    }

    fn write_footnotes<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        let left_padding = self.left_padding();

        for footnote in &self.footnotes {
            write!(
                writer,
                "{:padding$} {} ",
                "",
                '>'.style(config.styles.footnote_indicator),
                padding = left_padding
            )?;
            writeln!(writer, "{}", footnote)?;
        }

        Ok(())
    }

    fn write_footnotes_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        for footnote in &self.footnotes {
            write!(writer, "{} ", '>'.style(config.styles.footnote_indicator))?;
            writeln!(writer, "{}", footnote)?;
        }

        Ok(())
    }

    /// Writes this diagnostic to the given [`Write`]r using the specified [`Config`].
    pub fn write_to<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        self.write_header(writer, config)?;
        self.write_body(writer, config)?;
        self.write_footnotes(writer, config)?;

        writeln!(writer)?;
        Ok(())
    }

    /// Writes this diagnostic to the given [`Write`]r using the specified [`Config`]
    /// in compact mode.
    pub fn write_to_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        self.write_header_compact(writer, config)?;
        self.write_body_compact(writer, config)?;
        self.write_footnotes_compact(writer, config)?;

        writeln!(writer)?;
        Ok(())
    }

    /// Writes this diagnostic to `stderr` using the specified [`Config`].
    #[inline]
    pub fn eprint(&self, config: &Config) -> std::io::Result<()> {
        let mut eout = BufWriter::new(std::io::stderr());
        self.write_to(&mut eout, config)?;
        Ok(())
    }

    /// Writes this diagnostic to `stderr` using the specified [`Config`]
    /// in compact mode.
    #[inline]
    pub fn eprint_compact(&self, config: &Config) -> std::io::Result<()> {
        let mut eout = BufWriter::new(std::io::stderr());
        self.write_to_compact(&mut eout, config)?;
        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::test::diagnostic_snapshot;

    #[test]
    fn test_singleline() {
        const SAMPLE: &str = include_str!("../samples/sample2.rs");
        let src = Source::new(SAMPLE, Some("src/lib.rs"));
        let diagnostic = Diagnostic::new("error[E0072]: recursive type `List` has infinite size")
            .with_label(Label::new(53..66u32, ""))
            .with_label(Label::new(83..87u32, "recursive without indirection"))
            .with_footnote("error: could not compile `playground` (lib) due to previous error")
            .with_source(src);

        diagnostic_snapshot!(diagnostic);
    }
}
