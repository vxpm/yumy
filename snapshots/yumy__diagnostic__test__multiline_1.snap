---
source: src/diagnostic.rs
expression: clean
---
error[E0277]: `Rc<Mutex<i32>>` cannot be sent between threads safely
   @ [src/main.rs]
13 │ ┬ let handle = thread::spawn(move || {
   : │              ^^^^^^^^^^^^^ required by a bound introduced by this call
14 │ │     let mut num = counter.lock().unwrap();
15 │ │ 
16 │ │     *num += 1;
17 │ ┼ });
   : ╰─ `Rc<Mutex<i32>>` cannot be sent between threads safely
   > help: within `{closure@src/main.rs:11:36: 11:43}`, the trait `Send` is not implemented for `Rc<Mutex<i32>>`
   > note: required because it's used within this closure


